vscode-smart-contracts/src/commands.ts

```diff
@@ -15,10 +19,155 @@ export async function version() {
+  editor: vscode.TextEditor,
+  schemaSettings: SchemaSettings
+) {
+  if (!(await haveWasmTargetInstalled())) {
+    const response = await vscode.window.showInformationMessage(
+      "The needed wasm32-unknown-unknown rust target seems to be missing. Should it be installed?",
+      "Install",
+      "Abort"
+    );
+    if (response === "Abort") {
+      vscode.window.showErrorMessage(
+        "Unable to build because of missing wasm32-unknown-unknown target."
+      );
+      return;
     }
+    await installWasmTarget();
   }
+  const cwd = path.dirname(editor.document.uri.fsPath);
+  const workspaceFolder = vscode.workspace.getWorkspaceFolder(
+    editor.document.uri
+  );
+  const schemaArgs =
+    schemaSettings === "skip"
+      ? []
+      : schemaSettings === "embed"
+      ? ["--schema-embed"]
+      : ["--schema-base64-out", "-"];
+  const additionalArgs = config.getAdditionalBuildArgs();
+  const args = schemaArgs.concat(additionalArgs);
+  return vscode.tasks.executeTask(
+    await cargoConcordium.build(cwd, workspaceFolder, args)
+  );
+}
+
+/**
+ * Check the rustup list of installed targets for wasm32-unknown-unknown.
+ */
+async function haveWasmTargetInstalled() {
+  const out = await exec("rustup target list --installed");
+  return out.stdout.includes("wasm32-unknown-unknown");
+}
+
+/**
+ * Install the wasm32-unknown-unknown target using rustup.
+ * The returned promise resolves when the install task have ended.
+ */
+function installWasmTarget() {
+  const execution = new vscode.ProcessExecution("rustup", [
+    "target",
+    "install",
+    "wasm32-unknown-unknown",
+  ]);
+  const task = new vscode.Task(
+    { type: cargoConcordium.CONCORDIUM_TASK_TYPE, command: "install wasm" },
+    vscode.TaskScope.Workspace,
+    `Install WASM target`,
+    "Build smart contract",
+    execution
+  );
+  return executeAndAwaitTask(task);
+}
+
+/**
+ * Execute and await a task to end.
+ * @param task The task to execute.
+ * @returns Promise which resolves when the task have ended.
+ */
+function executeAndAwaitTask(task: vscode.Task) {
+  return new Promise<void>((resolve, reject) => {
+    vscode.tasks.onDidEndTask((event) => {
+      if (event.execution.task === task) {
+        resolve();
+      }
+    });
+    try {
+      vscode.tasks.executeTask(task);
+    } catch (e) {
+      reject(e);
+    }
+  });
+}
+
+/**
+ * Run 'cargo-concordium test' using the directory of the currently focused editor.
+ */
+export async function test(editor: vscode.TextEditor) {
+  if (!(await haveWasmTargetInstalled())) {
+    const response = await vscode.window.showInformationMessage(
+      "The needed wasm32-unknown-unknown rust target seems to be missing. Should it be installed?",
+      "Install",
+      "Abort"
+    );
+    if (response === "Abort") {
+      vscode.window.showErrorMessage(
+        "Unable to run tests because of missing wasm32-unknown-unknown target."
+      );
+      return;
+    }
+    await installWasmTarget();
+  }
```

Comment:
Nitpicking: Since you check (and prompt the user to install) the correct wasm target in the `test` and `build` functions, this part could be in a separate function so we don't duplicate code.

Line 150-163

Reply:
Right, so I did consider this, but I find the resulting code less readable. Mainly because of having to propagate the return early from an inner function.